### ISSUE TITLE
[Bug Report] Inconsistent Values When Converting Torch Tensor to TTNN Tensor with bfloat8_b Type

### ISSUE THREAD

**Issue by ngohoang34** [tenstorrent/tt-metal#14090]
**Describe the bug**
When creating a ttnn tensor with the `bfloat8_b` type from a Torch tensor containing random numbers between 0 and 1, the ttnn tensor does not match the Torch tensor values exactly, even though all numbers should have the same exponent (0). There are small differences in the printed values, which is unexpected.

**To Reproduce**
1. Create a Torch tensor with random numbers between 0 and 1, using the torch.bfloat16 data type.
2. Convert the Torch tensor to a ttnn tensor with the bfloat8_b type.
3. Print the values of both the ttnn tensor and the original Torch tensor.
4. Compare the printed results to observe differences.

```
npu_dtype = ttnn.bfloat8_b
cpu_dtype = torch.bfloat16
npu_layout = ttnn.TILE_LAYOUT

torch_input = torch.rand(input_shape, dtype=cpu_dtype, requires_grad=True) 

tt_input = ttnn.from_torch(torch_input, dtype=npu_dtype, layout=npu_layout, device=device)
tt_input_cpu = ttnn.to_torch(tt_input)

print(tt_input_cpu)
print(torch_input)
```

**Expected behavior**
The values in the ttnn tensor should exactly match the values in the original Torch tensor, as both tensors represent numbers between 0 and 1 with the same exponent (0).

**Screenshots**
Here are the printed results showing the differences:
```
torchTensor([[[[0.3438, 0.1016, 0.7734, 0.7188, 0.6328, 0.8594, 0.0156, 0.4531,
                0.2031, 0.8359, 0.3984, 0.8750, 0.0859, 0.5703, 0.9375, 0.7109,
                0.6875, 0.3047, 0.1562, 0.7734, 0.8438, 0.9453, 0.5156, 0.2266,
                0.1094, 0.8203, 0.4297, 0.9375, 0.7812, 0.2500, 0.9453],...
```
```
ttnnTensor([[[[0.3398, 0.0977, 0.7734, 0.7148, 0.6289, 0.8594, 0.0117, 0.4531,
           0.2031, 0.8359, 0.3945, 0.8750, 0.0859, 0.5664, 0.9336, 0.7070,
           0.6836, 0.3008, 0.1523, 0.7695, 0.8438, 0.9453, 0.5117, 0.2266,
           0.1094, 0.8164, 0.4297, 0.9375, 0.7773, 0.2461, 0.9414],...
```

**Please complete the following environment information:**
- OS: Ubuntu 22.04.4
- Version of software: [567c64b](https://github.com/tenstorrent/tt-metal/commit/567c64b0b83243695054c92054ff7cf4018e91dc)
- Device: n150

**Comment by ayerofieiev-tt** [comment]
@ngohoang34 , thank you! Let me take a look and follow up after the initial investigation this week.

**Comment by ngohoang34** [comment]
Update: There was a misunderstanding on my part. The random numbers should actually be between 1 and 2 to share the same exponent. I’ve updated the test accordingly, but there are still some differences between the two tensors. You can take a look at [this file](https://github.com/tenstorrent/tt-metal/blob/ngohoang34/test_bfp8_b/tests/ttnn/unit_tests/operations/test_bfp8_b.py) to help with debugging.

**Comment by jvasilje** [comment]
@ayerofieiev-tt This is a P0...? Any update?

Moreh team - is this a critical blocking bug?

**Comment by razorback3** [comment]
> [@ayerofieiev-tt](https://github.com/ayerofieiev-tt) This is a P0...? Any update?
> 
> Moreh team - is this a critical blocking bug?

Yes.
Currently, we are testing the bfp8 implementation of every ops and these kind of issues are making us hard to debug whether the precision loss is due to the implementation or bugs like this.

**Comment by zzigler-tt** [comment]
@ayerofieiev-tt Can you please have the team look into this item and advise next step or expected resolution timeline? TY

**Comment by ntarafdar** [comment]
I'll sync with @jaykru-tt and get an ETA on this.

**Comment by zzigler-tt** [comment]
@ntarafdar Hey there, requesting an update on this item, thanks.

**Comment by ntarafdar** [comment]
@zzigler-tt i think by end of week. @jaykru-tt Is going to try the update from the blackhole team tomorrow and rest of week work on this

**Comment by jaykru-tt** [comment]
Starting on this today!

**Comment by jaykru-tt** [comment]
@zzigler-tt This doesn't seem like a TM issue. I set breakpoints on the likely TM ops (untilize*, tilize*) and they aren't entered during this test. I think this is like a runtime to TTNN issue in `ttnn.to_device`.

**Comment by zzigler-tt** [comment]
@jaykru-tt please assign over and pull in the necessary folks to help remediate.

**Comment by TT-BrianLiu** [comment]
`torch.rand` for `float32` guarantees values in [0, 1) but if it's `bfloat16`, it can actually be [0, 1] (note the inclusive end). I think this is a bug on pytorch side: https://github.com/pytorch/pytorch/issues/96947

I think what's happening is otrch just converts to `bfloat16` after, so we get loss of precision when rounding up (as opposed to just truncating the mantissa which would result in even more error?).

If you run this code, you will get exact match:
```
npu_dtype = ttnn.bfloat8_b
cpu_dtype = torch.bfloat16
npu_layout = ttnn.TILE_LAYOUT

input_shape = [1, 1, 32, 32]
torch_input = torch.rand(input_shape, dtype=torch.bfloat16, requires_grad=True) + 1
torch_input = torch.where(torch_input < 2, torch_input, 1.5)

tt_input = ttnn.from_torch(torch_input, dtype=npu_dtype, layout=npu_layout, device=device)
tt_input_cpu = ttnn.to_torch(tt_input)

print("got: ", tt_input_cpu)
print("expected: ", torch_input)
```

**Comment by cmaryanTT** [comment]
@ngohoang34 - can you please comment on the findings above.

**Comment by ngohoang34** [comment]
@TT-BrianLiu thank you for your reply. I ran the code you provided but still noticed some differences:

got:  TorchTensor([[[[1.0781, **1.7656**, 1.0938,  ..., 1.9688, 1.1250, 1.7031],
               [1.2188, 1.5781, 1.2500,  ..., 1.7188, **1.6875**, 1.7812],
               [1.3750, 1.9375, 1.6719,  ..., 1.8906, **1.1875**, 1.2969],
               ...,
               [**1.6406**, 1.3125, 1.9375,  ..., 1.0938, 1.6406, **1.9219**],
               [1.5625, 1.5156, **1.9531**,  ..., 1.4844, 1.5156, 1.1406],
               [1.2031, 1.9062, 1.1719,  ..., 1.6562, 1.2969, 1.0781]]]])

expected:  tensor([[[[1.0781, **1.7578**, 1.0938,  ..., 1.9688, 1.1250, 1.7031],
          [1.2188, 1.5781, 1.2500,  ..., 1.7188, **1.6797**, 1.7812],
          [1.3750, 1.9375, 1.6719,  ..., 1.8906, **1.1797**, 1.2969],
          ...,
          [**1.6328**, 1.3125, 1.9375,  ..., 1.0938, 1.6406, **1.9141**],
          [1.5625, 1.5156, **1.9453**,  ..., 1.4844, 1.5156, 1.1406],
          [1.2031, 1.9062, 1.1719,  ..., 1.6562, 1.2969, 1.0781]]]],
       dtype=torch.bfloat16, grad_fn=<WhereBackward0>)

**Comment by TT-BrianLiu** [comment]
I can take a look

**Comment by TT-BrianLiu** [comment]
For BFP8, we also additionally store the hidden bit in the MSB of mantissa (which is always 1 for BF16 and is not stored). So we end up losing the LSB, which accounts for the difference (even if exponents are the exact same). I updated this slide here: 
<img width="1032" alt="Image" src="https://github.com/user-attachments/assets/fcf05f51-f4f5-479a-90b4-abd0fb2e0b15" />

**Comment by ngohoang34** [comment]
Thank you for your explanation. I'll close this issue
